specs/_features/eip7805/fork-choice.md

```diff
@@ -85,35 +164,111 @@ def get_forkchoice_store(anchor_state: BeaconState, anchor_block: BeaconBlock) -
         unrealized_justifications={anchor_root: justified_checkpoint},
-        # [New in EIP-7805]
-        unsatisfied_inclusion_list_blocks=set(),
+        unsatisfied_inclusion_list_blocks=set(),  # [New in EIP7805]
     )
 ```
 
+### New `get_inclusion_list_store`
+
+```python
+def get_inclusion_list_store() -> InclusionListStore:
+    # `cached_or_new_inclusion_list_store` is implementation and context dependent.
+    # It returns the cached `InclusionListStore`; if none exists,
+    # it initializes a new instance, caches it and returns it.
+    inclusion_list_store = cached_or_new_inclusion_list_store()
+
+    return inclusion_list_store
+```
+
+### New `store_inclusion_list`
+
+```python
+def store_inclusion_list(
+    store: InclusionListStore, inclusion_list: InclusionList, is_before_view_freeze_deadline: bool
+) -> None:
+    key = (inclusion_list.slot, inclusion_list.inclusion_list_committee_root)
+
+    inclusion_lists = store.inclusion_lists.setdefault(key, set())
+    equivocators = store.equivocators.setdefault(key, set())
+
+    # Ignore `inclusion_list` from equivocators.
+    if inclusion_list.validator_index in equivocators:
+        return
+
+    for stored_inclusion_list in inclusion_lists:
+        if stored_inclusion_list.validator_index != inclusion_list.validator_index:
+            continue
+
+        if stored_inclusion_list != inclusion_list:
+            equivocators.add(inclusion_list.validator_index)
+            inclusion_lists.remove(stored_inclusion_list)
+
+        # Whether it was an equivocation or not, we have processed this `inclusion_list`.
+        return
+
+    # Only store `inclusion_list` if it arrived before the view freeze deadline.
+    if is_before_view_freeze_deadline:
+        inclusion_lists.add(inclusion_list)
+```
+
+### New `get_inclusion_list_transactions`
+
+*Note*: `get_inclusion_list_transactions` returns a list of unique transactions
+from all valid and non-equivocating `InclusionList`s that were received in a
+timely manner on the p2p network for the given slot and for which the
+`inclusion_list_committee_root` in the `InclusionList` matches the one
+calculated based on the current state.
+
+```python
+def get_inclusion_list_transactions(
+    store: InclusionListStore, state: BeaconState, slot: Slot
+) -> Sequence[Transaction]:
+    inclusion_list_committee = get_inclusion_list_committee(state, slot)
+    inclusion_list_committee_root = hash_tree_root(inclusion_list_committee)
+    key = (slot, inclusion_list_committee_root)
+
+    inclusion_lists = store.inclusion_lists.setdefault(key, set())
+    equivocators = store.equivocators.setdefault(key, set())
+
+    inclusion_list_transactions = {
+        transaction
+        for inclusion_list in inclusion_lists
+        if inclusion_list.validator_index not in equivocators
+        for transaction in inclusion_list.transactions
+    }
+
+    return list(inclusion_list_transactions)
```

Comment:
This is implicitly deduping the list of transactions, right? I think we should make this a little more explicit, because I almost missed this myself. Maybe something like this:

```suggestion
    inclusion_list_transactions = [
        transaction
        for inclusion_list in inclusion_lists
        if inclusion_list.validator_index not in equivocators
        for transaction in inclusion_list.transactions
    ]

    # Deduplicate, order does not need to be preserved
    return list(set(inclusion_list_transactions))
```

Reply:
Well, the deduplication comes from the fact that `inclusion_lists` field of `InclusionListStore` has the value of `set`. Would adding a comment regarding this inside `get_inclusion_list_transactions` enhance readability?

Reply:
Hmm but what if there are multiple inclusion lists with the same transaction?

Eg you cannot do this:

```python
    return [
        transaction
        for inclusion_list in inclusion_lists
        if inclusion_list.validator_index not in equivocators
        for transaction in inclusion_list.transactions
    ]
```

Reply:
Ohh you're right. Let me fix it now. Great catch!

Reply:
Now I finally understood your first comment. Reflected the suggestion.

I think I'm being Pythonic a bit too much. Let me know if other parts of the spec have a room for readability improvement.